intake_esm/core.py

```diff
@@ -660,13 +676,16 @@ def _get_subset(df, require_all_on=None, **query):
 
     query = _normalize_query(query)
     for key, val in query.items():
-        if isinstance(val, (tuple, list)):
-            condition_i = np.zeros(len(df), dtype=bool)
-            for val_i in val:
+        condition_i = np.zeros(len(df), dtype=bool)
+        for val_i in val:
+            if isinstance(df[key].dtype, (np.object, pd.core.arrays.string_.StringDtype)):
+                if isinstance(val_i, Pattern):
+                    condition_i = condition_i | (df[key].str.contains(val_i, regex=True))
+                else:
+                    condition_i = condition_i | (df[key] == val_i)
+            else:
                 condition_i = condition_i | (df[key] == val_i)
-            condition = condition & condition_i
-        elif val is not None:
-            condition = condition & (df[key] == val)
+        condition = condition & condition_i
```

Comment:
@andersy005 You are doing a lot of `isinstance` checking here.  Do you need it?  Is seems that the `search()` method is just a "pass-through" interface to the `pandas.DataFrame[col].str.contains()` method.  If so, why not let Pandas do the error checking for you?  You could really simplify the code a lot and just tell people that:

```python
esm_datastore.search(key=val)
```

is equivalent to:

```python
esm_datastore.df[key].str.contains(val)
```

Are there any other parts of the API that are thin layers on top of Pandas?  If so, then we can pass off the error checking to Pandas.

Reply:
>  It seems that the search() method is just a "pass-through" interface to the pandas.DataFrame[col].str.contains() method

This would only be true if you had a pandas dataframe in which all column dtypes are strings. However, in most cases, the pandas dataframe has columns with a mix of dtypes. For instance, when  a user calls `esm_datastore.search(key1=val1, key2=val2)`, the keys here (`key1`, `key2` may refer to columns with different dtypes, and the `.str` accessor is available for string-like dtypes, and this is why I am doing the type checking.

Reply:
> If so, why not let Pandas do the error checking for you? 

Could you expand on this? I am asking because if you assume that all columns are of `string` dtype, you get the following error, and I don't see how intake-esm should handle this. 

```python
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'key1' :[1, 2], 'key2': ['foo', 'bar']})

In [3]: df
Out[3]:
   key1 key2
0     1  foo
1     2  bar

In [4]: df.key1.str
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-ba969f2734e5> in <module>
----> 1 df.key1.str

~/opt/miniconda3/envs/analysis/lib/python3.7/site-packages/pandas/core/generic.py in __getattr__(self, name)
   5268             or name in self._accessors
   5269         ):
-> 5270             return object.__getattribute__(self, name)
   5271         else:
   5272             if self._info_axis._can_hold_identifiers_and_holds_name(name):

~/opt/miniconda3/envs/analysis/lib/python3.7/site-packages/pandas/core/accessor.py in __get__(self, obj, cls)
    185             # we're accessing the attribute of the class, i.e., Dataset.geo
    186             return self._accessor
--> 187         accessor_obj = self._accessor(obj)
    188         # Replace the property with the accessor object. Inspired by:
    189         # http://www.pydanny.com/cached-property.html

~/opt/miniconda3/envs/analysis/lib/python3.7/site-packages/pandas/core/strings.py in __init__(self, data)
   2039
   2040     def __init__(self, data):
-> 2041         self._inferred_dtype = self._validate(data)
   2042         self._is_categorical = is_categorical_dtype(data)
   2043         self._is_string = data.dtype.name == "string"

~/opt/miniconda3/envs/analysis/lib/python3.7/site-packages/pandas/core/strings.py in _validate(data)
   2096
   2097         if inferred_dtype not in allowed_types:
-> 2098             raise AttributeError("Can only use .str accessor with string values!")
   2099         return inferred_dtype
   2100

AttributeError: Can only use .str accessor with string values!
```

Reply:
You are correct that this `search` method only maps to `.str.contains` if the `dtype` of the column is `string`, so my comment doesn't make sense.

My mind is just trying to figure out how to reduce the number of `if`-checks because there is only 1 case where you want to use `str.contains`, and all other cases are just `==` operations.  Maybe there isn't a way.

I guess you can mark this as resolved.

Reply:
> My mind is just trying to figure out how to reduce the number of `if`-checks because there is only 1 case where you want to use `str.contains`

you could do something like

```python
        for val_i in val:
            if isinstance(df[key].dtype, (np.object, pd.core.arrays.string_.StringDtype)) and isinstance(val_i, Pattern):
                new_condition = (df[key].str.contains(val_i, regex=True))
            else:
                new_condition = (df[key] == val_i)
            condition_i = condition_i | new_condition
```

Reply:
...Let's see.  You have 2 conditions that must be met to use `str.contains`, and if either of those 2 conditions are not met, then you just use `==`, correct?

So, you could improve readability of the code by avoiding the nested `if` block and doing something like:

```python
column_is_stringtype = isinstance(df[key].dtype, (np.object, pd.core.arrays.string_.StringDtype))
value_is_repattern = isinstance(val_i, Pattern)

if column_is_stringtype and value_is_repattern:
    cond = df[key].str.contains(val_i, regex=True)
else:
    cond = (df[key] == val_i)

condition_i = condition_i | cond
```